### PR TITLE
New audit log implementation

### PR DESCRIPTION
**Pull request by acuiuli**
### List of changes
- [feat(op-app): new audit use case](https://github.com/pagopa/io-fims/pull/64/commits/28592aea87e2949d1e40a27bc53c7c0c256e0f3d)

### Motivation and context
This pr adds the implementation to manage the event register required for legal reason as described in the FIMS DR

### Type of changes

- [ ] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [ ] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [x] No

### Other information

---

### CODE REVIEW

File: apps/op-app/src/use-cases/audit.ts
Comment (Garma00) [inline apps/op-app/src/use-cases/audit.ts:61]: Can't we parse this with zod to avoid the cast?

File: apps/op-app/src/use-cases/audit.ts
Comment (Garma00) [inline apps/op-app/src/use-cases/audit.ts:86]: Ditto

**Review by Garma00 — COMMENTED** [review]
I noticed that we don't have an `env.example` do we plan to add it so we can list the env variables needed?

File: apps/op-app/src/adapters/oidc/middleware.ts
Comment (acuiuli) [reply apps/op-app/src/adapters/oidc/middleware.ts:11]: This is the middleware as specified by the node-oidc-provider documentation [here](https://github.com/panva/node-oidc-provider/blob/main/docs/README.md#pre--and-post-middlewares)

File: packages/io-fims-common/src/domain/audit-event.ts
Comment (acuiuli) [reply packages/io-fims-common/src/domain/audit-event.ts:17]: The audit event schema will be overriden once the other pr is merged and all the validation are correct there

File: apps/op-app/src/use-cases/audit.ts
Comment (acuiuli) [reply apps/op-app/src/use-cases/audit.ts:64]: Yes but I'm waiting until we merge the other pr with the correct schema

File: apps/op-app/src/use-cases/audit.ts
Comment (acuiuli) [reply apps/op-app/src/use-cases/audit.ts:89]: Yes but I'm waiting until we merge the other pr with the correct schema

File: apps/op-app/src/adapters/express/routes/interaction.ts
Comment (acuiuli) [reply apps/op-app/src/adapters/express/routes/interaction.ts:183]: This is not a concern of this pr, so I will resolve it

File: apps/op-app/src/adapters/redis/event.ts
Comment (acuiuli) [reply apps/op-app/src/adapters/redis/event.ts:28]: This function is always called inside a try catch [here](https://github.com/pagopa/io-fims/blob/6bac0448e84ec6ddd7b270a85dd2d234561e1c1c/apps/op-app/src/domain/session.ts#L82)

File: apps/op-app/src/adapters/redis/event.ts
Comment (acuiuli) [reply apps/op-app/src/adapters/redis/event.ts:37]: This function is always called inside a try catch [here](https://github.com/pagopa/io-fims/blob/6bac0448e84ec6ddd7b270a85dd2d234561e1c1c/apps/op-app/src/domain/session.ts#L77)

File: apps/op-app/src/adapters/storage/event-client.ts
Comment (acuiuli) [reply apps/op-app/src/adapters/storage/event-client.ts:20]: This function is always called inside a try catch [here](https://github.com/pagopa/io-fims/blob/6bac0448e84ec6ddd7b270a85dd2d234561e1c1c/apps/op-app/src/domain/storage.ts#L12)

File: apps/op-app/src/adapters/env.ts
Comment (lucacavallaro) [inline apps/op-app/src/adapters/env.ts:7]:
```suggestion
```

this variable has been removed from `main` since it is no longer used

File: apps/op-app/src/adapters/express/routes/interaction.ts
Comment (lucacavallaro) [inline apps/op-app/src/adapters/express/routes/interaction.ts:182]: we should raise an error if `ipAddress` is empty (because `""` is not valid/accettable value for an `ipAddress`)

File: apps/op-app/src/use-cases/audit.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/audit.ts:38]: > A use case is a piece of business logic that represents a single task that the system needs to perform. The use case encapsulates the rules and logic required to perform the task, and defines the inputs and outputs required for the operation.

https://nanosoft.co.za/blog/post/clean-architecture-use-cases

----

so they are not "services" classes that group related function, they must represent only one action

**Review by lucacavallaro — CHANGES_REQUESTED** [review]

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:22]: I think `event` it is not specific enough. You should replace it with something like `auditEvent` (and its variations `AUDIT_EVENT`, `AuditEvent`)

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:58]: this function should be splitted in two half (`execute` can parse the payload and call the right private function based on the value of `type`)

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:30]:
```suggestion
    idToken: z.string().min(1),
```

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:66]: Are there specific reasons that led you to throw `Error` objects here? if not, you can omit `new Error` and just pass the `message` as the second argument to `assert.ok`. By default, it throws `AssertionError` objects.

```suggestion
      assert.ok(clientId, "The clientId is undefined");
```

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:77]: What if `blobName` is `undefined`?

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:93]: prefer destructuring

```suggestion
      const { sessionId, requestParams } = params;
```

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:101]: what if `userData?.fiscalCode` is `undefined` ?

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (lucacavallaro) [inline apps/op-app/src/use-cases/send-event-messge.ts:102]: The name `redisEvent` leaks implementation details. At `use-case level` the code don't should know the existence of `redis`.

**Review by lucacavallaro — DISMISSED** [review]

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (acuiuli) [reply apps/op-app/src/use-cases/send-event-messge.ts:66]: No reason, just a typo that I forgot to remove after a refactor. I will remove it

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (acuiuli) [reply apps/op-app/src/use-cases/send-event-messge.ts]: I added a check to make sure that userData is not undefined. The fiscalCode field cannot be undefined.

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (acuiuli) [reply apps/op-app/src/use-cases/send-event-messge.ts]: I removed this part with the methods refactor

File: apps/op-app/src/use-cases/send-event-messge.ts
Comment (acuiuli) [reply apps/op-app/src/use-cases/send-event-messge.ts]: I added a check to avoid that event is undefined. The blobName field cannot be undefined.

**Review by lucacavallaro — APPROVED** [review]

**Review by lucacavallaro — APPROVED** [review]

**Pull request merged**
